### PR TITLE
Feat/study closing: Add option to close a study

### PR DESCRIPTION
**Pull request by hig-dev**
This PR adds the feature to close a study. It is not possible to re-open a study. No new participants will be able to enroll in a closed study.

- [x] Database: Add "is_closed" column to study table
- [x] Database: Add policy to restrict joining a closed study
- [x] App: Do not show closed studies in study selection
- [x] App: Show dialog if user tries to join a closed study
- [x] Designer: Use study status to show state of study (running, draft, closed)
- [x] Designer: Adapt study duplicate function
- [x] Designer: Add feature to close study using modal actions
- [x] Designer: Add banner to recruit page for closed study

### CODE REVIEW

**Review by johannesvedder — CHANGES_REQUESTED** [review]

File: database/studyu-schema.sql [hunk 1]
```diff
@@ -835,6 +836,19 @@ CREATE POLICY "Editors can see subjects from their studies" ON public.study_subj
...
+
+CREATE POLICY "Joining a closed study should not be possible" ON public.study_subject
+AS RESTRICTIVE
+FOR INSERT
+WITH CHECK (NOT EXISTS (
+    SELECT 1
+    FROM public.study
+    WHERE study.id = study_subject.study_id
+    AND study.is_closed
+));
```
[the remark this answers is not in the document]
Comment (hig-dev) [reply database/studyu-schema.sql:851]: Good catch. Yes, the `upsert` operation in `StudySubject.save()` is problematic.

I would suggest to replace the upsert with explicit insert or update operations.

[the remark this answers is not in the document]
Comment (hig-dev) [reply database/studyu-schema.sql:851]: See: https://github.com/hpi-studyu/studyu/pull/601/commits/62243d8d19227d85a4dcf40bbdfe0fc4e0a42f52

File: core/lib/src/models/tables/study.dart [hunk 2]
```diff
@@ -54,6 +54,8 @@ class Study extends SupabaseObjectFunctions<Study> implements Comparable<Study>
   @JsonKey(name: 'icon_name')
   late String iconName = 'accountHeart';
   late bool published = false;
+  @JsonKey(name: 'is_closed', defaultValue: false)
+  bool isClosed = false;
```
Comment (johannesvedder) [inline core/lib/src/models/tables/study.dart:58]: rename to `closed`

File: core/lib/src/models/tables/study.dart [hunk 3]
```diff
@@ -170,7 +172,7 @@ class Study extends SupabaseObjectFunctions<Study> implements Comparable<Study>
   static Future<ExtractionResult<Study>> publishedPublicStudies() async {
     ExtractionResult<Study> result;
     try {
-      final response = await env.client.from(tableName).select().eq('participation', 'open');
+      final response = await env.client.from(tableName).select().eq('participation', 'open').eq("is_closed", false);
```
Comment (johannesvedder) [inline core/lib/src/models/tables/study.dart:175]: .

File: core/lib/src/models/tables/study.dart [hunk 4]
```diff
@@ -240,8 +245,6 @@ class Study extends SupabaseObjectFunctions<Study> implements Comparable<Study>
 
   bool get isDraft => status == StudyStatus.draft;
   bool get isRunning => status == StudyStatus.running;
-  // TODO: missing flag to indicate that study is completed & enrollment closed
-  bool get isClosed => false;
```
Comment (johannesvedder) [inline core/lib/src/models/tables/study.dart:248]: .

File: database/migration/migrate-close_study.sql [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+
+CREATE POLICY "Joining a closed study should not be possible" ON public.study_subject
+    AS RESTRICTIVE
+    FOR INSERT
+    WITH CHECK (NOT EXISTS (
+    SELECT 1
+    FROM public.study
+    WHERE study.id = study_subject.study_id
+      AND study.is_closed
+));
```
Comment (johannesvedder) [inline database/migration/migrate-close_study.sql:12]: rename file

File: database/studyu-schema.sql [hunk 6]
```diff
@@ -81,6 +81,7 @@ CREATE TABLE public.study (
     description text NOT NULL,
     icon_name text NOT NULL,
     published boolean DEFAULT false NOT NULL,
+    is_closed boolean DEFAULT false NOT NULL,
```
Comment (johannesvedder) [inline database/studyu-schema.sql:84]: .

File: database/studyu-schema.sql (hunk 1 above)
[the remark this answers is not in the document]
Comment (johannesvedder) [reply database/studyu-schema.sql:851]: test this again

File: designer_v2/lib/domain/study.dart [hunk 7]
```diff
@@ -111,6 +114,7 @@ extension StudyDuplicateX on Study {
     copy.title = (copy.title ?? '').withDuplicateLabel();
     copy.userId = userId;
     copy.published = false;
+    copy.isClosed = false;
```
Comment (johannesvedder) [inline designer_v2/lib/domain/study.dart:117]: .

File: designer_v2/lib/localization/app_en.arb [hunk 8]
```diff
@@ -68,7 +68,7 @@
   "study_status_running": "Live",
   "study_status_running_description": "This study is currently in progress.",
   "study_status_closed": "Closed",
-  "study_status_closed_description": "This study has been completed.",
+  "study_status_closed_description": "This study has been completed.\nNo new participants will be able to enroll.",
```
Comment (johannesvedder) [inline designer_v2/lib/localization/app_en.arb:71]: ..will not be able..

File: designer_v2/lib/localization/app_en.arb [hunk 9]
```diff
@@ -119,6 +119,9 @@
   "study_settings_publish_results_tooltip": "Make your anonymized study results & data available in the study registry. \nOther researchers & clinicians will be able to access, export and \nanalyze the results from your study (the Analyze page will be available). \n This will automatically publish your study design to the registry.",
   "action_button_study_launch": "Launch",
   "notification_study_deleted": "Study was deleted",
+  "notification_study_closed": "Study was closed",
+  "dialog_study_close_title": "Close participation?",
+  "dialog_study_close_description": "Are you sure that you want to close participation for this study? No new participants will be able to enroll.",
```
Comment (johannesvedder) [inline designer_v2/lib/localization/app_en.arb:124]: .

File: designer_v2/lib/localization/app_en.arb [hunk 10]
```diff
@@ -241,6 +244,8 @@
 
   "banner_study_readonly_title": "This study cannot be edited.",
   "banner_study_readonly_description": "You can only make changes to studies where you are an owner or collaborator. Studies that have been launched cannot be changed by anyone.",
+  "banner_study_closed_title": "This study is closed.",
+  "banner_study_closed_description": "No new participants will be able to enroll in this study.",
```
Comment (johannesvedder) [inline designer_v2/lib/localization/app_en.arb:248]: .

**Review by johannesvedder — CHANGES_REQUESTED** [review]

File: app/lib/screens/study/onboarding/study_selection.dart [hunk 11]
```diff
@@ -143,6 +159,10 @@ class _StudySelectionScreenState extends State<StudySelectionScreen> {
                             child: StudyTile.fromStudy(
                               study: study,
                               onTap: () async {
+                                if (study.isClosed) {
+                                  await showStudyClosedDialog(context);
+                                  return;
+                                }
                                 await navigateToStudyOverview(context, study);
```
Comment (johannesvedder) [inline app/lib/screens/study/onboarding/study_selection.dart:166]: are closed studies shown in study selection? should we make them invisible instead?

**Review by johannesvedder — CHANGES_REQUESTED** [review]

**Review by johannesvedder — APPROVED** [review]

**Pull request merged**